pandas/io/excel.py

```diff
@@ -66,27 +69,164 @@ def get_writer(engine_name):
         raise ValueError("No Excel writer '%s'" % engine_name)
 
 
+class BaseFile(object):
+    """Base class for excel readers
+
+    A file class can be initialized even if the engine is not installed.
+    If the engine is not installed, io_class and _open_workbook are
+    type(None) and None respectively. When attempting to use open_workbook
+    while _open_workbook is None, the relevant ImportError is raised. When
+    calling is_type or is_ext, False is retunred when the reader is not
+    installed.
+    """
+    # added in order to deal with both Open Document Format and MS Excel
+    # fil types, see PR #9070
+
+    def __init__(self, engine, extensions):
+        """Set the engine name, and extension. If the engine is not installed,
+        io_class and _open_workbook are type(None) and None respectively,
+        but the ImportError is ingored at the init stage.
+        """
+        self.engine = engine
+        self.extensions = extensions
+        self.io_class = type(None)
+        self._open_workbook = None
+        # fail silently if engine is not presented on init, only raise when
+        # attempting to use it when opening a workbook
+        try:
+            self.load_engine()
```

Comment:
What I am trying to do here is to be able to create a file reader class that does not fail if its dependencies are not installed. The logic for deriving the excel reader type now iterates over the supported types, regardless of the availability of their dependencies. In that way the file type can still be determined for certain cases, and give the user a more clear error message when a dependency is missing. The switching logic is in ```ExcelFile()```, starting from line nr [313](https://github.com/pydata/pandas/pull/9070/files#diff-0153a977179cfc73a2645cf0b7cd1f7fR313).